sync/_databricks.py

```diff
@@ -887,10 +1031,33 @@ def _get_job_cluster(tasks: List[dict], job_clusters: list) -> Response[dict]:
     return Response(error=DatabricksError(message="Not all tasks use the same cluster"))
 
 
+def _filter_run_tasks(
+    all_tasks: List[dict],
+    exclude_tasks: Union[Collection[str], None] = None,
+    project_id: str = None,
+):
+    tasks = [
+        task for task in all_tasks if not exclude_tasks or task["task_key"] not in exclude_tasks
+    ]
+    if project_id:
+        project_tasks = [
+            t
+            for t in tasks
+            if t.get("new_cluster", {}).get("custom_tags", {}).get("sync:project-id") == project_id
+        ]
```

Comment:
Looks like this could potentially get us tasks that run on different clusters. And what about the tasks that run on a job cluster?

What if we create a mapping of clusters to tasks here with only the non-excluded tasks. Filter the dict by project ID if we have one. Error if we end up with an empty dict or one with more than 1 item.

Maybe we move this all into `_get_run_cluster_id` and rename it - it'll return that 1 dict item or error. for the job clusters we'll have to pull it down in a separate API transaction so we may want a `_get_cluster_report` that takes a cluster object and have the current `_get_cluster_report` call it after getting a cluster if we still need that interface.

Reply:
> Looks like this could potentially get us tasks that run on different clusters.

This is true, though in practice this is enforced when we call `_get_run_cluster_id` in order to get the cluster info.

> Maybe we move this all into _get_run_cluster_id and rename it - it'll return that 1 dict item or error.

Yeah, I think this is a good idea :+1:

> And what about the tasks that run on a job cluster?

Do you mean when we're applying a recommendation to a Job Definition? As far as I'm aware, we don't distinguish between e.g. All-Purpose and Jobs Compute clusters when creating a submission 🤔